src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+                    // \\?\
+                    path = &path[2..];
+                    if path.starts_with(b"UNC\\") {
+                        // \\?\UNC\server\share
+                        path = &path[4..];
+                        let (server, share) = match parse_two_comps(path, is_verbatim_sep) {
+                            Some((server, share)) => (u8_slice_as_os_str(server),
+                                                      u8_slice_as_os_str(share)),
+                            None => (u8_slice_as_os_str(path),
+                                     u8_slice_as_os_str(&[])),
+                        };
+                        return Some(VerbatimUNC(server, share));
+                    } else {
+                        // \\?\path
+                        let idx = path.position_elem(&b'\\');
+                        if idx == Some(2) && path[1] == b':' {
+                            let c = path[0];
+                            if c.is_ascii() && (c as char).is_alphabetic() {
+                                // \\?\C:\ path
+                                let slice = u8_slice_as_os_str(&path[0..1]);
+                                return Some(VerbatimDisk(slice));
+                            }
+                        }
+                        let slice = &path[.. idx.unwrap_or(path.len())];
+                        return Some(Verbatim(u8_slice_as_os_str(slice)));
+                    }
+                } else if path.starts_with(b".\\") {
+                    // \\.\path
+                    path = &path[2..];
+                    let slice = &path[.. path.position_elem(&b'\\').unwrap_or(path.len())];
+                    return Some(DeviceNS(u8_slice_as_os_str(slice)));
+                }
+                match parse_two_comps(path, is_sep) {
+                    Some((server, share)) if server.len() > 0 && share.len() > 0 => {
+                        // \\server\share
+                        return Some(UNC(u8_slice_as_os_str(server),
+                                        u8_slice_as_os_str(share)));
+                    }
+                    _ => ()
+                }
+            } else if path.len() > 1 && path[1] == b':' {
+                // C:
+                let c = path[0];
+                if c.is_ascii() && (c as char).is_alphabetic() {
+                    return Some(Disk(u8_slice_as_os_str(&path[0..1])));
+                }
+            }
+            return None;
+        }
+
+        fn parse_two_comps(mut path: &[u8], f: fn(u8) -> bool) -> Option<(&[u8], &[u8])> {
+            let first = match path.iter().position(|x| f(*x)) {
+                None => return None,
+                Some(x) => &path[.. x]
+            };
+            path = &path[(first.len()+1)..];
+            let idx = path.iter().position(|x| f(*x));
+            let second = &path[.. idx.unwrap_or(path.len())];
+            Some((first, second))
+        }
+    }
+
+    #[derive(Copy, Clone, PartialEq)]
```

Comment:
Not sure if I'll see something down below, but are paths like `C:\foo` and `c:\foo` equivalent? If so, does this derived implementation handle that?